pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -951,16 +951,17 @@ func (l *LibvirtDomainManager) generateConverterContext(vmi *v1.VirtualMachineIn
 	var efiConf *converter.EFIConfiguration
 	if vmi.IsBootloaderEFI() {
 		secureBoot := vmi.Spec.Domain.Firmware.Bootloader.EFI.SecureBoot == nil || *vmi.Spec.Domain.Firmware.Bootloader.EFI.SecureBoot
+		enrolledKeys := vmi.Spec.Domain.Firmware.Bootloader.EFI.EnrolledKeys == nil || *vmi.Spec.Domain.Firmware.Bootloader.EFI.EnrolledKeys
```

Comment:
The matrix of possibilities here are troublesome. This would allow for all three options described in [secureboot docs](https://libvirt.org/kbase/secureboot.html).

If we are going this route we should probably switch to firmware auto-selection which might require a bigger refactor (e.g: let libvirt handle the keys) and have a well defined use cases that we want to support (we will have problems with secureboot in different archs and even in different OS versions).

@andreabolognani, I should have recorded that 1h+ discussion. Would you mind taking a look and see if my note here is valid or I misremember ?

Reply:
Yes I would agree that using firmware autoselection seems to be a better idea. My workaround in the original issue was already trying to use that.

Reply:
And what problems you've faced when using firmware autoselection?

Reply:
You cannot specify loader and nvram along with firmware autoselect

Reply:
That was true in the past, but it's not the case anymore.

The change was introduced with [this series](https://lists.libvirt.org/archives/list/devel@lists.libvirt.org/thread/22MYBS3LUZIBS2ZJ4XUGVYY6VQDYQQB6/), which is part of libvirt 9.2.0. Some follow-up fixes were included in libvirt 9.7.0.

These days, the only constraint is that you can't provide information that disagree with each other, for example

```xml
<os firmware='efi'>
  <firmware>
    <feature enabled='yes' name='secure-boot'/>
  </firmware>
  <loader format='qcow2'>/usr/share/edk2/ovmf/OVMF_CODE_4M.qcow2</loader>
</os>
```

This configuration will be rejected because libvirt knows that the specific build of edk2 referenced in the `<loader>` element doesn't have Secure Boot support, which is explicitly requested in the `<firmware>` element.

The following configuration, on the other hand, is perfectly valid:

```xml
<os firmware='efi'>
  <firmware>
    <feature enabled='no' name='enrolled-keys'/>
    <feature enabled='yes' name='secure-boot'/>
  </firmware>
  <loader type='pflash' format='qcow2'>/usr/share/edk2/ovmf/OVMF_CODE_4M.secboot.qcow2</loader>
  <nvram template='/usr/share/edk2/ovmf/OVMF_VARS_4M.qcow2' format='qcow2'/>
</os>
```

In fact, if you provide just

```xml
<os firmware='efi'>
  <firmware>
    <feature enabled='no' name='enrolled-keys'/>
    <feature enabled='yes' name='secure-boot'/>
  </firmware>
</os>
```

or just 

```xml
<os>
  <loader type='pflash' format='qcow2'>/usr/share/edk2/ovmf/OVMF_CODE_4M.secboot.qcow2</loader>
  <nvram template='/usr/share/edk2/ovmf/OVMF_VARS_4M.qcow2' format='qcow2'/>
</os>
```

libvirt will fill in the blanks and the end result will be full snippet seen above in both cases.

Reply:
There's one important caveat that I failed to mention. The mechanism I just described only works on the condition that the user-provided paths are **the same** as the ones contained in the corresponding firmware descriptor. This is how libvirt can match the two.

Right now KubeVirt hardcodes legacy paths such as `/usr/share/OVMF/OVMF_CODE.fd`, which can't be matched with any firmware descriptor. So none of the above will work, and you're right that in that specific context using the two mechanisms together is not possible.

Incidentally, this also means that KubeVirt does not benefit from the improvements made somewhat recently in the firmware area, such as the switch to 4MB builds and the use of qcow2 format, which results in significant memory usage improvements for aarch64 VMs.